R/parallelization.R

```diff
@@ -52,58 +102,19 @@
 #' mirai::daemons(0)
 #' }
 #'
-#' All daemons automatically terminate when your session ends and the connection
-#' drops. Hence you do not need to explicitly terminate daemons in this instance,
-#' although it is still good practice to do so.
+#' All daemons automatically terminate when your session ends. You do not need
+#' to explicitly terminate daemons in this instance, although it is still good
+#' practice to do so.
 #'
 #' Note: it should always be for the user to set daemons. If you are using
 #' parallel map within a package, do not make any [mirai::daemons()] calls
 #' within the package. This helps prevent inadvertently spawning too many
 #' daemons if functions are used recursively within each other.
 #'
-#' # Crating a function
-#'
-#' [carrier::crate()] provides a systematic way of making the function `.f`
-#' self-contained so that it can be readily shared with other processes.
-#'
-#' Crating ensures that everything needed by the function is serialized along
-#' with it, but not other objects which happen to be in the function's enclosing
-#' environment. This helps to prevent inadvertently shipping large data objects
-#' to daemons, where they are not needed.
-#'
-#' Any non-package function supplied to `.f` will be automatically crated. When
-#' this happens, a confirmation along with the crate size is printed to the
-#' console. Package functions are not crated as these are already
-#' self-contained.
-#'
-#' If your function `.f()` contains free variables, for example it references
-#' other local functions in its body, then explicitly [carrier::crate()] your
-#' function supplying these variables to its `...` argument. This ensures that
-#' these objects are available to `.f()` when it is executed in a parallel
-#' process.
-#'
-#' Examples:
-#' \preformatted{
-#' # package functions are not auto-crated:
-#' map(1:3, stats::runif, .parallel = TRUE)
-#'
-#' # other functions (incl. anonymous functions) are auto-crated:
-#' mtcars |> map_dbl(function(...) sum(...), .parallel = TRUE)
-#'
-#' # explicitly crate a function to include other objects required by it:
-#' fun <- function(x) \{x + x \%\% 2 \}
-#' map(1:3, carrier::crate(function(x) x + fun(x), fun = fun), .parallel = TRUE)
-#' }
-#'
-#' For details on further options, see [carrier::crate].
-#'
-#' # Further documentation
+#' @references
 #'
 #' \pkg{purrr}'s parallelization is powered by \CRANpkg{mirai}. See the
 #' [mirai website](https://mirai.r-lib.org/) for more details.
 #'
-#' Crating is provided by the \CRANpkg{carrier} package. See the
-#' [carrier readme](https://github.com/r-lib/carrier) for more details.
-#'
-#' @name parallelization
-NULL
+#' @export
+parallelize <- carrier::crate
```

Comment:
```suggestion
parallelize <- fucntion(.fn, ...) {
   check_installed(c("carrier", "mirai"))   
   carrier::crate(.fn. ...)
}
```

There are a couple of advantages: most importantly this avoids literally copying the crate function in at build-time (which 99% of the time is fine, but when it's not causes extremely confusing bugs), but also allows you to move carrier back to suggests (which I think make sense because carrier and mirai should have the same status). 

It's not strictly necessary to check for mirai here but I think it gives a better user experience.

Reply:
I also wonder if it's still worth printing the size of the crated function?

Reply:
> There are a couple of advantages: most importantly this avoids literally copying the crate function in at build-time (which 99% of the time is fine, but when it's not causes extremely confusing bugs), but also allows you to move carrier back to suggests (which I think make sense because carrier and mirai should have the same status).

I'm aware of this, and I did initially try this way.

However `carrier::crate()` is a bit special in this case as it requires a freshly-defined function. All my attempts at getting around this have caused at least one failure in the tests.

My suggestion - either leave it as is, or we vendor the function into purrr. I've tested that vendoring works, and the definition's fairly minimal. The advantage of this is that we can also tailor the error messages to mention `in_parallel()` rather than `crate()`.

Inviting @lionel-'s view. I think either way works, as it's a very limited scope package where we wouldn't be making changes frequently, if at all.

Reply:
I guess we would want `enexpr()` instead of `substitute()` so callers can inject stuff in the crated function? Though implicit injection will cause surprising behaviour if the parallel function uses tidyeval. It's likely better to require explicit injection.

Reply:
> This should work:
> 
> ```r
> inject(carrier::crate(!!substitute(fn), ...)
> ```
> 
> I've opened [r-lib/carrier#14](https://github.com/r-lib/carrier/issues/14) regarding error messages.

This still causes one test to fail: https://github.com/tidyverse/purrr/blob/main/tests/testthat/test-parallel.R#L101-L114. Same using `enexpr()`.

With the error:
```r
Error (test-parallel.R:105:3): primitive dispatch correctly
Error: object 'method' not found
Backtrace:
     ▆
  1. ├─testthat::expect_identical(...) at test-parallel.R:105:3
  2. │ └─testthat::quasi_label(enquo(object), label, arg = "object")
  3. │   └─rlang::eval_bare(expr, quo_get_env(quo))
  4. ├─purrr::map(...)
  5. │ └─purrr:::map_("list", .x, .f, ..., .progress = .progress) at purrr/R/map.R:164:3
  6. │   └─carrier::is_crate(.f) at purrr/R/map.R:202:3
  7. └─purrr::in_parallel(function(x) as.character(x), as.character.test_class = method)
  8.   ├─rlang::inject(carrier::crate(!!substitute(.f), ...), env = environment()) at purrr/R/parallelization.R:121:3
  9.   └─carrier::crate(function(x) as.character(x), ...)
 10.     └─carrier:::locally(!!!dots, .env = env)
 11.       └─rlang::eval_bare(dots[[i]], .env)
```

Reply:
Oh I think I found the solution:
```r
inject(carrier::crate(!!substitute(.f), !!!list(...)))
```
Thanks again!

Reply:
@hadley's suggestions have now been implemented in [8b48804](https://github.com/tidyverse/purrr/pull/1185/commits/8b4880480ba87fb05963229bfff2754b2d05fa4a).